pkg/env/certificateRefresher_test.go

```diff
@@ -0,0 +1,283 @@
+			sleep:    10 * time.Millisecond,
+			managerFactory: func(controller *gomock.Controller) keyvault.Manager {
+				manager := mock_keyvault.NewMockManager(controller)
+				manager.EXPECT().GetCertificateSecret(gomock.Any(), testCertName).Return(key1, certs1, nil)
+				return manager
+			},
+			wantKey:  key1,
+			wantCert: certs1[0],
+			wantErr:  nil,
+		},
+		{
+			name:     "test refresh certificate, pull exactly twice, first on start, second on refresh",
+			interval: 50 * time.Millisecond,
+			sleep:    60 * time.Millisecond,
+			managerFactory: func(controller *gomock.Controller) keyvault.Manager {
+				manager := mock_keyvault.NewMockManager(mockController)
+				gomock.InOrder(
+					manager.EXPECT().GetCertificateSecret(gomock.Any(), testCertName).Return(key1, certs1, nil),
+					manager.EXPECT().GetCertificateSecret(gomock.Any(), testCertName).Return(key2, certs2, nil),
+				)
+				return manager
+			},
+			wantKey:  key2,
+			wantCert: certs2[0],
+			wantErr:  nil,
+		},
+		{
+			name:     "test initial error, pull exactly once with an error",
+			interval: 1 * time.Minute,
+			sleep:    20 * time.Millisecond,
+			managerFactory: func(controller *gomock.Controller) keyvault.Manager {
+				manager := mock_keyvault.NewMockManager(mockController)
+				manager.EXPECT().GetCertificateSecret(gomock.Any(), testCertName).Return(nil, nil, cannotPull)
+				return manager
+			},
+			wantKey:  nil,
+			wantCert: nil,
+			wantErr:  cannotPull,
+		},
+		{
+			name:     "test refresh error, pull exactly twice, first on start, second time with an error",
+			interval: 50 * time.Millisecond,
+			sleep:    60 * time.Millisecond,
+			managerFactory: func(controller *gomock.Controller) keyvault.Manager {
+				manager := mock_keyvault.NewMockManager(controller)
+				gomock.InOrder(
+					manager.EXPECT().GetCertificateSecret(gomock.Any(), testCertName).Return(key1, certs1, nil),
+					manager.EXPECT().GetCertificateSecret(gomock.Any(), testCertName).Return(nil, nil, cannotPull),
+				)
+				return manager
+			},
+			wantKey:  key1,
+			wantCert: certs1[0],
+			wantErr:  nil,
+		},
+		{
+			name:     "test refresh, pull exactly 10 times",
+			interval: 10 * time.Millisecond,
+			sleep:    100 * time.Millisecond,
+			managerFactory: func(controller *gomock.Controller) keyvault.Manager {
+				manager := mock_keyvault.NewMockManager(controller)
+				manager.EXPECT().GetCertificateSecret(gomock.Any(), testCertName).Return(key1, certs1, nil).Times(10)
+				return manager
+			},
+			wantKey:  key1,
+			wantCert: certs1[0],
+			wantErr:  nil,
+		},
+	}
+
+	for _, test := range tt {
+		t.Run(test.name, func(t *testing.T) {
+			fmt.Println(test.name)
+			stop := make(chan struct{})
+			refreshing := newCertificateRefresher(
+				context.Background(),
+				logrus.NewEntry(logrus.StandardLogger()),
+				test.interval,
+				test.managerFactory(mockController),
+				testCertName,
+				stop,
+			)
+
+			err := refreshing.Start(context.Background())
+			defer close(stop)
```

Comment:
Nit: I would move it directly under `stop := make(chan struct{})`. It is a common pattern for `defer` calls to go straight after creating something.